rfd/headless-login.md

```diff
@@ -0,0 +1,367 @@
+
+### Headless authentication overview
+
+The headless authentication flow is shown below:
+
+```mermaid
+sequenceDiagram
+    participant Local Machine
+    participant Remote Machine
+    participant Teleport Proxy
+    participant Teleport Auth
+
+    Note over Remote Machine: tsh --headless ssh user@node01
+    Remote Machine->>+Teleport Proxy: initiate headless login
+    Teleport Proxy->>Teleport Auth: save request with new uuid
+    Note over Teleport Auth: Request: {id, user, publicKey, ...}
+    Teleport Auth->>Teleport Proxy: request ID
+    Teleport Proxy->>-Remote Machine: request ID
+    Note over Remote Machine: print command + URL
+
+    Remote Machine->>+Teleport Proxy: request signed user certs<br/>for request ID
+    Note over Local Machine: user copies command / URL<br/>to local terminal / browser
+    Local Machine->>+Teleport Proxy: initiate headless request approval
+    Teleport Proxy->>Local Machine: MFA challenge
+    Note over Local Machine: user taps YubiKey
+    Local Machine->>Teleport Proxy: signed MFA challenge response
+    Teleport Proxy-->>-Local Machine: success/failure
+
+    Teleport Proxy->>Teleport Auth: Request signed user certificates
+    Teleport Auth->>Teleport Proxy: Issue signed user certificates<br/>for the client's public key 
+    Teleport Proxy->>-Remote Machine: User certificates
+    Note over Remote Machine: Connect to user@node01
+```
+
+This flow can be broken down into three parts: headless login initiation, local authentication, and certificate retrieval.
+
+#### Headless login initiation
+
+First, the client initiates a headless login. In this stage, the client provides request details including the Teleport username, client public key, and the command being requested (e.g. `tsh ssh user@host`). The request will begin in the pending state, to be later approved/denied by the user locally.
+
+```go
+type HeadlessLoginRequest struct {
+    // User is a teleport username
+    User string `json:"user"`
+    // PubKey is a public key user wishes to sign in
+    PubKey []byte `json:"pub_key"`
+    // Command is the client command being requested
+    Command string
+    // State is the request state (pending, approved, or denied)
+    State string
+}
+```
+
+These request details are saved on the Auth server in a `HeadlessLoginRequest` object, with the backend prefix `/headless_login_requests/`. They will be saved with a 1 minute TTL, by which point the user should have completed the authentication flow.
+
+In response, the client receives a request ID generated by the server, which it will use to generate a client command or URL to begin local authentication: `tsh headless approve <request_id>` and `https://proxy.example.com/headless/<request_id>/approve`.
```

Comment:
This is interesting idea. I wonder, would it make sense to build it as extension of existing access requests? I.e. reuse the same access request object and approval workflow. Technically, we'd be extending access requests from the scenario "I have role X + I want role Y" to also supporting the scenario "I have no roles (== not logged in) + I want role X" (i.e. "zero to one").

This might be just a semantics question but I'm just wondering if having yet another "approve" command might be a little confusing. WDYT? @Joerger @jakule @klizhentas

Reply:
I like the idea. I'd potentially make the implementation simpler, but wouldn't that also introduce some weird UX flows? We would need to distinguish between headless and access requests. Sounds like a place with higher chances to introduce regression as `GetAll()` would return all types and then we would need to filter them.